### ISSUE TITLE
NonStrict arguments should be allowed to occur relevantly in the type

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#1602]
The following (very simple) example is currently rejected by Agda:
```agda
id : ..{A : Set} → A → A
id x = x
```
> Variable A is declared irrelevant, so it cannot be used here

In Agda.Syntax.Common:
```haskell
-- | Irrelevant function arguments may appear non-strictly in the codomain type.
irrToNonStrict :: Relevance -> Relevance
irrToNonStrict Irrelevant = NonStrict
-- irrToNonStrict NonStrict  = Relevant -- TODO: is that what we want (OR: NonStrict)  -- better be more conservative
irrToNonStrict rel        = rel
```
I think the second clause should be uncommented.

Original issue reported on code.google.com by `jesper.c...@gmail.com` on 6 Jul 2015 at 12:09

**Comment by jespercockx** [comment]
The current situation is quite crazy: my example `id` in the original post is accepted by Agda, EXCEPT when you enable --experimental-irrelevance, then you get an error. This is almost certainly not what we want.

**Comment by jespercockx** [comment]
Aha, I see @UlfNorell fixed this issue under the guise of #2128, but the case for --experimental-irrelevance was left unchanged. The problem is that when we combine both features (turning irrelevant into nonstrict and nonstrict into relevant) then we can eventually get from irrelevant to relevant, which is maybe not desirable.

**Comment by jespercockx** [comment]
I thought about it for a bit and I don't immediately see what could go wrong if we do both (famous last words, I know). The testsuite also passes without problems (perhaps unsurprisingly since --experimental-irrelevance isn't used very much). Since this option literally has *experimental* in the name, I propose I push my change and we'll see what happens.

**Comment by andreasabel** [comment]
@UlfNorell hijacked my `NonStrict` modality for run-time irrelevance.  However, I intended `NonStrict` arguments as shape-irrelevant.  I have not thought about what the original example
```agda
id : ..(A : Set) -> A -> A
id A x = x
``` 
should mean in terms of shape-irrelevance, since `id` does not necessarily construct a type, thus, formation of `.(A : Set) -> id A` is illegal for other reasons.

My intention for shape-irrelevance in types is e.g. to allow level-irrelevance in terms.
```
Set : ..(l : Level) -> Set (lsuc l)

id : .(l : Level) (A : Set l) -> A -> A
id l A x = x
```

**Comment by andreasabel** [comment]
Should be revisited once runtime-erasure is implemented, see #2513.

**Comment by jespercockx** [comment]
Now runtime-irrelevance is implemented, so we should revisit this issue.
